File: lua-class-loader/classfile.lua
-- ClassFile class for ClassLoader.
-- Ported from C++ to Lua.
--
-- Copyright 2015 Sam Saint-Pettersen
--
-- Released under the MIT/X11 License.
-- Please see LICENSE file.

ClassFile = {}
ClassFile.__index = ClassFile

function ClassFile.create()

	local self = setmetatable({}, ClassFile)

	self.magic = 0
	self.minor_version = 0
	self.major_version = 0
	self.constant_pool_count = 0
	self.constant_pool = {}
	self.cp_size = {}
	self.access_flags = 0
	self.this_class = 0
	self.super_class = 0
	self.start_byte = 0
	self.interfaces_count = 0
	self.interfaces = {}
	self.fields_count = 0
	self.fields = {}
	self.methods_count = 0
	self.methods = {}
	self.attributes_count = 0
	self.attributes = {}
	self.bytecodes_count = 0
	self.bytecodes = {}
	self.tags = { 'Null', 'Utf8', 'Integer', 'Float', 'Long', 'Double' }
	table.insert(self.tags, 'Class')
	table.insert(self.tags, 'String') 
	table.insert(self.tags, 'Fieldref')
	table.insert(self.tags, 'Methodref')
	table.insert(self.tags, 'InterfaceMethodref')
	table.insert(self.tags, 'NameAndType')

	return self
end

function ClassFile:setMagicNumber(magicNum)
	self.magic = magicNum
end

function ClassFile:getMagicNumber()
	return self.magic
end

function ClassFile:checkMagicNumber()
	isMagic = false
	if self.magic == 0xCAFEBABE then 
		isMagic = true
	end
	return isMagic
end

function ClassFile:setMinorVersion(minor)
	self.minor_version = minor
end

function ClassFile:getMinorVersion()
	return self.minor_version
end

function ClassFile:setMajorVersion(major)
	self.major_version = major
end

function ClassFile:getMajorVersion()
	return self.major_version
end

function ClassFile:setCosntantPoolCount(cpCount)
	self.constant_pool_count = cpCount
end

function ClassFile:getCPCOUNT()
	return self.constant_pool_count
end

function ClassFile:pushToConstantPool(constValPair)
	table.insert(self.constant_pool, constValPair)
end

function ClassFile:setCPSIZE(tagSize, _type)
	print('+ ' .. tagSize .. '( ' .. _type .. ')') --!
	table.insert(self.cp_size, tagSize)
end

function ClassFile:getCPSIZE()
	f_cpsize = 0
	for size in self.cpsize do
		f_cpsize = f_cpsize + size
	end
	return f_cpsize
end

function ClassFile:setAccessFlags(accessFlags)
	self.access_flags = accessFlags
end

function ClassFile:getAccessFlags()
	return self.access_flags
end

function ClassFile:setThisClass(thisClass)
	self.this_class = thisClass
end

function ClassFile:getThisClass()
	return self.this_class
end

function ClassFile:setSuperClass(superClass)
	self.super_class = superClass
end

function ClassFile:getSuperClass()
	return self.super_class
end

function ClassFile:setStartByte(start)
	self.start_byte = start
end

function ClassFile:getStartByte()
	return self.start_byte
end

function ClassFile:setInterfacesCount(interfacesCount)
	self.interfaces_count = interfacesCount
end

function ClassFile:getInterfacesCount()
	return self.interfaces_count
end

function ClassFile:pushInterface(interface)
	table.insert(self.interfaces, interface)
end

function ClassFile:pullInterfaces()
	return self.interfaces
end

function ClassFile:setFieldsCount(fieldsCount)
	self.fields_count = fieldsCount
end

function ClassFile:getFieldsCount()
	return self.fields_count
end

function ClassFile:pushField(field)
	table.insert(self.fields, field)
end

function ClassFile:pullFields()
	return self.fields
end

function ClassFile:setMethodsCount(methodsCount)
	self.methods_count = methodsCount
end

function ClassFile:getMethodsCount()
	return self.methods_count
end

function ClassFile:pushMethod(method)
	table.insert(self.methods, method)
end

function ClassFile:pullMethods()
	return self.methods
end

function ClassFile:setAttributesCount(attribsCount)
	self.attributes_count = attribsCount
end

function ClassFile:pushAttibute(attribute)
	table.insert(self.attributes, attribute)
end

function ClassFile:pullAttributes()
	return self.attributes
end

function ClassFile:setBytecodesCount(bytcodesCount)
	self.bytecodes_count = bytcodesCount
end

function ClassFile:getBytecodesCount()
	return self.bytecodes_count
end

function ClassFile:pushBytecode(bytecode)
	table.insert(self.bytecodes, bytecode)
end

function ClassFile:pullBytecodes()
	return self.bytecodes
end

function ClassFile:getTag(index)
	local rtag = 'Null'
	if index ~= 0 or index ~= 2 or index <= 12 or index > 0 then
		rtag = tags[index]
	end
	return rtag
end

File: go-class-loader/classfile.go
/*

ClassFile class for ClassLoader.

Copyright 2015 Sam Saint-Pettersen

Released under the MIT/X11 License.
Please see LICENSE file.

*/

package main

import "fmt"

type ClassFile struct {
	magic uint32
	minor_version uint16
	major_version uint16
	constant_pool_count uint16
	constant_pool []ConstantValuePair
	cp_size []uint16
	access_flags uint16
	this_class uint16
	super_class uint16
	start_byte byte
	interfaces_count uint16
	interfaces []uint16
	fields_count uint16
	fields []uint16
	methods_count uint16
	methods []uint16
	attributes_count uint16
	attributes []uint16
	bytecodes_count uint16
	bytecodes []uint16
	tags []string
}

type ConstantValuePair struct {
	constVal1 string
	constVal2 string
}

func (cf *ClassFile) new() {
	cf.tags = []string { "Null", "Utf8", "Integer", "Float", "Long", "Double" }
	cf.tags[6]  = "Class"
	cf.tags[7]  = "String"
	cf.tags[8]  = "Fieldref"
	cf.tags[9]  = "Methodref"
	cf.tags[10] = "InterfaceMethodref"
	cf.tags[11] = "NameAndType"
}
func (cf *ClassFile) setMagicNumber(magicNum uint32)  {
	cf.magic = magicNum
}
func (cf *ClassFile) getMagicNumber() (uint32) {
	return cf.magic
}
func (cf *ClassFile) checkMagicNumber() (bool) {
	var isMagic bool = false
	if(cf.magic == 0xCAFEBABE) {
		isMagic = true
	}
	return isMagic
}
func (cf *ClassFile) setMinorVersion(minorVer uint16) {
	cf.minor_version = minorVer
}
func (cf *ClassFile) getMinorVersion() (uint16) {
	return cf.minor_version
}
func (cf *ClassFile) setMajorVersion(majorVer uint16) {
	cf.major_version = majorVer
}
func (cf *ClassFile) getMajorVersion() (uint16) {
	return cf.major_version
}
func (cf *ClassFile) setConstantPoolCount(cpCount uint16) {
	cf.constant_pool_count = cpCount
}
func (cf *ClassFile) getCPCOUNT() (uint16) {
	return cf.constant_pool_count
}
func (cf *ClassFile) pushToConstantPool(constValPair ConstantValuePair) {
	cf.constant_pool = append(cf.constant_pool, constValPair)
}
func (cf *ClassFile) setCPSIZE(tagSize uint16, _type string) {
	fmt.Println(fmt.Sprintf("+ %d (%v)", tagSize, _type))
	cf.cp_size = append(cf.cp_size, tagSize)
}
func (cf *ClassFile) getCPSIZE() (uint16) {
	var f_cpsize uint16 = 0
	for i := 0; i < len(cf.cp_size); i++ {
		f_cpsize += cf.cp_size[i]
	}
	return f_cpsize
}
func (cf *ClassFile) setAccessFlags(accessFlags uint16) {
	cf.access_flags = accessFlags
}
func (cf *ClassFile) getAccessFlags() (uint16) {
	return cf.access_flags
}
func (cf *ClassFile) setThisClass(thisClass uint16) {
	cf.this_class = thisClass
}
func (cf *ClassFile) getThisClass() (uint16) {
	return cf.this_class
}
func (cf *ClassFile) setSuperClass(superClass uint16) {
	cf.super_class = superClass
}
func (cf *ClassFile) getSuperClass() (uint16) {
	return cf.super_class
}
func (cf *ClassFile) setStartByte(start byte) {
	cf.start_byte = start
}
func (cf *ClassFile) getStartByte() (byte) {
	return cf.start_byte
}
func (cf *ClassFile) setInterfacesCount(interfacesCount uint16) {
	cf.interfaces_count = interfacesCount
}
func (cf *ClassFile) getInterfacesCount() (uint16) {
	return cf.interfaces_count
}
func (cf *ClassFile) pushInterface(_interface uint16) {
	cf.interfaces = append(cf.interfaces, _interface)
}
func (cf *ClassFile) pullInterfaces() ([]uint16) {
	return cf.interfaces
}
func (cf *ClassFile) setFieldsCount(fieldsCount uint16) {
	cf.fields_count = fieldsCount
}
func (cf *ClassFile) getFieldsCount() (uint16) {
	return cf.fields_count
}
func (cf *ClassFile) pushField(field uint16) {
	cf.fields = append(cf.fields, field)
}
func (cf *ClassFile) pullFields() ([]uint16) {
	return cf.fields
}
func (cf *ClassFile) setMethodsCount(methodsCount uint16) {
	cf.methods_count = methodsCount
}
func (cf *ClassFile) getMethodsCount() (uint16) {
	return cf.methods_count
}
func (cf *ClassFile) pushMethod(method uint16) {
	cf.methods = append(cf.methods, method)
}
func (cf *ClassFile) pullMethods() ([]uint16) {
	return cf.methods
}
func (cf *ClassFile) setAttributesCount(attribsCount uint16) {
	cf.attributes_count = attribsCount
}
func (cf *ClassFile) getAttributesCount() (uint16) {
	return cf.attributes_count
}
func (cf *ClassFile) pushAttribute(attribute uint16) {
	cf.attributes = append(cf.attributes, attribute)
}
func (cf *ClassFile) pullAttributes() ([]uint16) {
	return cf.attributes
}
func (cf *ClassFile) setBytecodesCount(bytecodesCount uint16) {
	cf.bytecodes_count = bytecodesCount
}
func (cf *ClassFile) getBytecodesCount() (uint16) {
	return cf.bytecodes_count
}
func (cf *ClassFile) pushBytecode(bytecode uint16) {
	cf.bytecodes = append(cf.bytecodes, bytecode)
}
func (cf *ClassFile) pullBytecodes() ([]uint16) {
	return cf.bytecodes
}
func (cf *ClassFile) getTag(index uint16) (string) {
	var rtag string = "Null"
	if(index != 0 || index != 2 || index <= 12 || index > 0) {
		rtag = cf.tags[index]
	}
	return rtag
}
